### PR TITLE
Cde violin plot

### PR DESCRIPTION
**Pull request by edlu77**
#627

Density plot portion of the visualization is functional. Still need to add the box plots in the center.

### CODE REVIEW

File: libs/cde-visualization/src/lib/components/violin/violin.component.html [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<span class="mat-caption violin-label">
+  Violin
```
Comment (bherr2) [inline libs/cde-visualization/src/lib/components/violin/violin.component.html:2]: Update to label and tooltip match spec should be "Violin Graph"
image

**Review by bherr2 — DISMISSED** [review]

File: libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts [hunk 2]
```diff
@@ -241,6 +252,26 @@ export class CdeVisualizationComponent {
...
+      return filtered.sort((a, b) => b.count - a.count);
+    },
+    { equal: emptyArrayEquals },
+  );
+
+  /** Computed distances between nodes */
+  readonly distances = computed(() => this.computeDistances(), { equal: emptyArrayEquals });
+
+  /** Data for the histogram visualization */
+  readonly data = computed(() => this.computeData(), { equal: emptyArrayEquals });
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts:270]: Add `protected`

File: libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts [hunk 3]
```diff
@@ -241,6 +252,26 @@ export class CdeVisualizationComponent {
...
+
+  /** Colors for the histogram visualization */
+  readonly colors = computed(() => this.computeColors(), { equal: emptyArrayEquals });
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts:273]: Add `protected`

File: libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts [hunk 4]
```diff
@@ -281,4 +312,53 @@ export class CdeVisualizationComponent {
...
+    const selection = new Set(this.cellTypesSelection());
+    if (selection.size === 0) {
+      return [];
+    }
+
+    return this.computeDistances().filter(({ type }) => selection.has(type));
+  }
+
+  /** Compute colors for the violin visualization */
+  computeColors(): string[] {
```
[the remark this answers is not in the document]
Comment (axdanbol) [reply libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts:349]: Rename to computeFilteredColors

File: libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts (hunk 3 above)
Comment (axdanbol) [inline libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts:273]: Rename to filteredColors

File: libs/cde-visualization/src/lib/components/histogram/histogram.component.ts [hunk 5]
```diff
@@ -20,28 +20,23 @@ import {
...
 import * as HISTOGRAM_SPEC from './histogram.vl.json';
 
-/** Interface for representing the distance entry */
-interface DistanceEntry {
-  /** Type of the entry */
-  type: string;
-  /** Distance value of the entry */
-  distance: number;
+interface UpdateColorData {
+  type: CellTypeEntry;
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/components/histogram/histogram.component.ts:38]: Rename to `entry`

File: libs/cde-visualization/src/lib/components/histogram/histogram.component.ts [hunk 6]
```diff
@@ -142,24 +137,17 @@ const DYNAMIC_COLOR_RANGE = Array(DYNAMIC_COLOR_RANGE_LENGTH)
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
 export class HistogramComponent {
-  /** List of nodes used in the histogram */
-  readonly nodes = input.required<NodeEntry[]>();
+  /** Data for the violin visualization */
+  readonly computedData = input.required<DistanceEntry[]>();
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/components/histogram/histogram.component.ts:141]: Rename to `data`

File: libs/cde-visualization/src/lib/components/histogram/histogram.component.ts [hunk 7]
```diff
@@ -142,24 +137,17 @@ const DYNAMIC_COLOR_RANGE = Array(DYNAMIC_COLOR_RANGE_LENGTH)
...
 
-  /** Key used to target specific node properties */
-  readonly nodeTargetKey = input.required<NodeTargetKey>();
+  /** Colors for the violin visualization */
+  readonly computedColors = input.required<string[]>();
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/components/histogram/histogram.component.ts:144]: Rename to colors

File: libs/cde-visualization/src/lib/components/histogram/histogram.component.ts [hunk 8]
```diff
@@ -261,64 +228,17 @@ export class HistogramComponent {
...
-  private computeData(): DistanceEntry[] {
-    const selection = new Set(this.cellTypesSelection());
-    if (selection.size === 0) {
-      return [];
-    }
-
-    return this.distances().filter(({ type }) => selection.has(type));
-  }
+  /** Color for the total cell type */
+  protected readonly totalCellTypeColor = signal<Rgb>([0, 0, 0], { equal: colorEquals });
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/components/histogram/histogram.component.ts:238]: Move up to to before methods (where it was before)

File: libs/cde-visualization/src/lib/components/violin/violin.component.ts [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+})
+export class ViolinComponent {
+  /** Tooltip position configuration */
+  readonly tooltipPosition = TOOLTIP_POSITION_RIGHT_SIDE;
+
+  /** State indicating whether the info panel is open */
+  infoOpen = false;
+
+  /** Data for the violin visualization */
+  readonly computedData = input.required<DistanceEntry[]>();
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/components/violin/violin.component.ts:127]: Rename to `data`

File: libs/cde-visualization/src/lib/components/violin/violin.component.ts [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+
+  /** Colors for the violin visualization */
+  readonly computedColors = input.required<string[]>();
```
Comment (axdanbol) [inline libs/cde-visualization/src/lib/components/violin/violin.component.ts:130]: Rename to `colors`

File: libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts [hunk 11]
```diff
@@ -281,4 +312,53 @@ export class CdeVisualizationComponent {
...
+    return this.filteredCellTypes()
+      .sort((a, b) => (a.name < b.name ? -1 : a.name === b.name ? 0 : 1))
+      .map(({ color }) => rgbToHex(color));
+  }
+
+  /** Update the color of a specific cell type entry */
+  updateColor(entry: CellTypeEntry, color: Rgb): void {
```
[the remark this answers is not in the document]
Comment (edlu77) [reply libs/cde-visualization/src/lib/cde-visualization/cde-visualization.component.ts:356]: This is used by the template so it can't be private

**Review by axdanbol — APPROVED** [review]

**Pull request merged**